_drafts/device-logging.md

```diff
@@ -0,0 +1,295 @@
+---
+title: Trade-offs and considerations for logging on embedded devices
+description: "Trade-offs and considerations for logging on embedded hardware"
+author: victorlai
+tags: [android, debugging, logging]
+---
+
+Logging on embedded devices comes with a host of interesting, unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there is still a lot of literature and learnings from logging on the server that we can also apply on the client side.
```

Comment:
nit: the use of interesting, unique challenges reads weird to me.
```suggestion
Logging on embedded devices comes with a host of interesting and unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there is still a lot of literature and learnings from logging on the server that we can also apply on the client side.
```

Reply:
Yeah it's intentionally more stream of conscious writing than formal essay. But and works a bit better anyways.